controllers/argocd/statefulset.go

```diff
@@ -441,13 +441,56 @@ func getArgoControllerContainerEnv(cr *argoprojv1a1.ArgoCD) []corev1.EnvVar {
 	return env
 }
 
-func (r *ReconcileArgoCD) reconcileApplicationControllerStatefulSet(cr *argoprojv1a1.ArgoCD, useTLSForRedis bool) error {
+func (r *ReconcileArgoCD) getArgoControllerReplicaCount(cr *argoprojv1a1.ArgoCD) int32 {
 	var replicas int32 = common.ArgocdApplicationControllerDefaultReplicas
 
-	if cr.Spec.Controller.Sharding.Replicas != 0 && cr.Spec.Controller.Sharding.Enabled {
+	if cr.Spec.Controller.Sharding.EnableDynamicScaling {
+		if cr.Spec.Controller.Sharding.MinShards < 1 {
+			log.Info("Minimum number of shards cannot be less than 1. Setting default value to 1")
+			cr.Spec.Controller.Sharding.MinShards = 1
+		}
+
+		if cr.Spec.Controller.Sharding.MaxShards < cr.Spec.Controller.Sharding.MinShards {
+			log.Info("Maximum number of shards cannot be less than minimum number of shards. Setting maximum shards same as minimum shards")
+		}
+
+		clusterSecrets, err := r.getClusterSecrets(cr)
+		if err != nil {
+			log.Info("Error retreiving cluster secrets for ArgoCD instance %s", cr.Name)
+		}
+
+		clustersOnEachShard := cr.Spec.Controller.Sharding.ClustersOnEachShard
+		if clustersOnEachShard < 1 {
+			log.Info("clustersOnEachShard cannot be less than 1. Defaulting to 1.")
```

Comment:
Similar comment. Add validation logic.

Reply:
Specifying both, max & min no. of shards at the same time as specifying the no. of clusters per shard can lead to a situation where one has to override the other

in this case it looks like we would prioritize staying within the max-min window over honoring no. of clusters per shard 

Is this intentional behavior, and if so should we document it somewhere?

Reply:
@jaideepr97 Yes, this is the intentional behavior. We would want to keep the number of replicas of application controller between the MinShards and MaxShards window. The clustersPerShard helps in deciding whether we should increase the number of replicas or not.

Reply:
@ishitasequeira how does this interface with the actual distribution algorithms that we are developing for cluster reallocation? For example, `clustersPerShard` is being used in a pretty simple way right now, but if in the future we have load based balacing in place and argo-cd wants to place 3 small clusters on one shard and 1 big cluster on another shard, that would go against whatever is defined in `clustersPerShard` by the user

Maybe we should at least make it clear to the user that whatever they configure in `clustersPerShard` won't always be honored

Reply:
@jaideepr97, this interface takes care of. distributing the clusters across multiple shards evenly for now. This is more of an intermediate step till ArgoCD starts considering load on a cluster as a factor to distribute shards. Considering load on each cluster is a bit of a complex thing in current scenario for ArgoCD, thus we decided to go step by step to achieve dynamic scaling of shards. 

As far as not honoring `clustersPerShard` goes, we would not honor the value in `clustersPerShard` even if the number of shards goes beyond the `minShards - maxShards` range. I can add this as a note in the docs.